test/core/event_engine/posix/dns_server.cc

```diff
@@ -0,0 +1,349 @@
+                     reinterpret_cast<const sockaddr*>(&query.client_addr),
+                     sizeof(query.client_addr));
+  if (sent < 0) {
+    return absl::ErrnoToStatus(errno, "Sending response");
+  }
+  return absl::OkStatus();
+}
+
+void DnsServer::SetIPv4Response(absl::Span<const uint8_t> ipv4_address) {
+  CHECK_EQ(ipv4_address.size(), 4u);
+  grpc_core::MutexLock lock(&mu_);
+  for (const auto& question : questions_) {
+    auto status = Respond(question, ipv4_address);
+    LOG_IF(FATAL, !status.ok()) << status;
+  }
+  questions_.clear();
+  ipv4_address_.emplace();
+  std::copy(ipv4_address.begin(), ipv4_address.end(), ipv4_address_->begin());
+}
+
+void DnsServer::ServerLoop(int sockfd) {
+  running_.Notify();
+  std::array<uint8_t, 2048> buffer;
+  sockaddr_in client_addr;
+  socklen_t client_len = sizeof(client_addr);
+  absl::Cleanup server_cleanup = [&, sockfd]() {
+    LOG(INFO) << "DNS server shutdown: " << done_.HasBeenNotified();
+    close(sockfd);
+  };
+
+  while (!done_.HasBeenNotified()) {
+    auto received_bytes = recvfrom(sockfd, buffer.data(), buffer.size(), 0,
+                                   (struct sockaddr*)&client_addr, &client_len);
+    if (received_bytes < 0 && (errno == EAGAIN || errno == EWOULDBLOCK)) {
+      // Don't wait
+      absl::SleepFor(absl::Milliseconds(5));
+      continue;
+    }
+    if (received_bytes < 0) {
+      LOG(FATAL) << absl::ErrnoToStatus(errno, "Reading from socket");
+      return;
+    }
+    auto query =
+        ParseQuestion(absl::Span<const uint8_t>(buffer).first(received_bytes));
+    if (!query.ok()) {
+      LOG(FATAL) << query.status();
+    }
+    LOG(INFO) << "Received question " << query->id << " for domain "
+              << query->qname;
+    query->client_addr = client_addr;
+    {
+      grpc_core::MutexLock lock(&mu_);
+      if (ipv4_address_.has_value()) {
+        auto response = Respond(*query, *ipv4_address_);
+        LOG_IF(FATAL, !response.ok()) << response;
+      } else {
+        questions_.emplace_back(std::move(query).value());
+        cond_.SignalAll();
+      }
+    }
+  }
+}
+
+absl::StatusOr<DnsServer> DnsServer::Start(int port) {
+  bool success = false;
+  int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+  if (sockfd > 0) {
+    int flags = fcntl(sockfd, F_GETFL, 0);
+    if (flags >= 0) {
+      success = fcntl(sockfd, F_SETFL, flags | O_NONBLOCK) == 0;
+    }
+  }
+  if (!success) {
+    return absl::ErrnoToStatus(errno, "Error creating socket");
+  }
+  sockaddr_in server_addr;
+  server_addr.sin_family = AF_INET;
+  server_addr.sin_port = htons(port);
+  server_addr.sin_addr.s_addr = INADDR_ANY;
+  if (bind(sockfd, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
+    auto status = absl::ErrnoToStatus(errno, "Error binding socket");
+    close(sockfd);
+    return status;
+  }
+  LOG(INFO) << "DNS server listening on port " << port;
+  return absl::StatusOr<DnsServer>(absl::in_place, port, sockfd);
```

Comment:
You might just be able to say this:

```
return DnsServer(port, sockfd);
```

Not sure if that would trigger problems with the type needing to be copyable.